File: Objective-C/TouchID/Controller/ViewController.m
//
//  ViewController.m
//  TouchID
//
//  Created by John Lima on 20/04/17.
//  Copyright © 2017 John Lima. All rights reserved.
//

#import "ViewController.h"
#import "Global.h"

@interface ViewController ()

@end

@implementation ViewController

#pragma mark - View LifeCycle
- (void)viewDidLoad {
    [super viewDidLoad];
    
}

-(void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
    [self.textField resignFirstResponder];
}

#pragma mark - Actions
-(IBAction)startTouchId:(UIButton *)sender {
    [self useTouchId];
}

-(void)useTouchId {

    LAContext *context = [[LAContext alloc] init];
    NSError *error = nil;
    
    if ([context canEvaluatePolicy:LAPolicyDeviceOwnerAuthenticationWithBiometrics error:&error]) {
        [context evaluatePolicy:LAPolicyDeviceOwnerAuthenticationWithBiometrics localizedReason:MESSAGE_PUT_FINGER_ON_SENSOR reply:^(BOOL success, NSError * _Nullable error) {
            dispatch_async(dispatch_get_main_queue(), ^{
                if (success) {
                    self.textField.text = MESSAGE_SECRET_MESSAGE;
                }else {
                    NSLog(@"Error: %@ - %@", MESSAGE_WRONG_FINGER, error.localizedDescription);
                }
            });
        }];
    }else {
        NSLog(@"%@", MESSAGE_UNSUPPORTED_DEVICE);
    }
}

@end

File: Swift/TouchID/ViewController.swift
//
//  ViewController.swift
//  TouchID
//
//  Created by John Lima on 4/21/15.
//  Copyright (c) 2015 John Lima. All rights reserved.
//

import UIKit
import LocalAuthentication

class ViewController: UIViewController {

    // MARK: - Properties
    @IBOutlet fileprivate weak var textField: UITextField!
    
    // MARK: - View LifeCycle
    override func viewDidLoad() {
        super.viewDidLoad()
        
    }

    // MARK: - Actions
    @IBAction fileprivate func startTouchId(sender: UIButton) {
        useTouchID()
    }
    
    fileprivate func useTouchID() {
        
        let authContext = LAContext()
        let authPolicy = LAPolicy.deviceOwnerAuthenticationWithBiometrics
        
        var authError: NSErrorPointer? = nil
        if (authContext.canEvaluatePolicy(authPolicy, error: authError!)) {
            
            authContext.evaluatePolicy(authPolicy, localizedReason: "Coloque o dedo no sensor biométrico do seu device", reply: {(success: Bool, error: NSError?) in
                
                DispatchQueue.main.async(execute: { () -> Void in
                    if (success) {
                        self.textField.text = "Isto é uma mensagem secreta"
                    }else {
                        print("Dedo errado")
                    }
                })
            } as! (Bool, Error?) -> Void)
            
        }else {
            print("Seu dispositivo não suporta o Touch ID")
        }
    }

}

